src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,926 @@
+ordIntR x y = compare (toSymmetricRelation x) (toSymmetricRelation y)
+
+instance Applicative Graph where
+    pure  = vertex
+    (<*>) = apR
+
+apR :: Graph (a -> b) -> Graph a -> Graph b
+apR f x = bindR f (<$> x)
+{-# INLINE apR #-}
+
+instance Monad Graph where
+    return = pure
+    (>>=)  = bindR
+
+bindR :: Graph a -> (a -> Graph b) -> Graph b
+bindR g f = foldg empty f overlay connect g
+{-# INLINE [0] bindR #-}
+
+instance Alternative Graph where
+    empty = empty
+    (<|>) = overlay
+
+instance MonadPlus Graph where
+    mzero = empty
+    mplus = overlay
+
+-- Help GHC with type inference (direct use of 'coerce' does not compile).
+coerce0 :: G.Graph a -> Graph a 
+coerce0 = coerce
+
+-- Help GHC with type inference (direct use of 'coerce' does not
+-- compile).
+coerce1 :: (a -> G.Graph b) -> (a -> Graph b)
+coerce1 = coerce
+
+-- Help GHC with type inference (direct use of 'coerce' does not
+-- compile).
+coerce2 :: (a -> a -> G.Graph b) -> (a -> a -> Graph b)
+coerce2 = coerce
+
+-- | Construct the /empty graph/.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- 'isEmpty'     empty == True
+-- 'hasVertex' x empty == False
+-- 'vertexCount' empty == 0
+-- 'edgeCount'   empty == 0
+-- 'size'        empty == 1
+-- @
+empty :: Graph a
+empty = coerce0 G.empty
+{-# INLINE empty #-}
+
+-- | Construct the graph comprising /a single isolated vertex/. 
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- 'isEmpty'     (vertex x) == False
+-- 'hasVertex' x (vertex x) == True
+-- 'vertexCount' (vertex x) == 1
+-- 'edgeCount'   (vertex x) == 0
+-- 'size'        (vertex x) == 1
+-- @
+vertex :: a -> Graph a
+vertex = coerce1 G.vertex
+{-# INLINE vertex #-}
+
+-- | Construct the graph comprising /a single edge/.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- edge x y               == 'connect' ('vertex' x) ('vertex' y)
+-- edge x y               == 'edge' y x
+-- edge x y               == 'edges' [(x,y), (y,x)]
+-- 'hasEdge' x y (edge x y) == True
+-- 'edgeCount'   (edge x y) == 1
+-- 'vertexCount' (edge 1 1) == 1
+-- 'vertexCount' (edge 1 2) == 2
+-- @
+edge :: a -> a -> Graph a
+edge = coerce2 G.edge
+
+-- | /Overlay/ two graphs. This is a
+-- commutative, associative and idempotent operation with the identity 'empty'.
+-- Complexity: /O(1)/ time and memory, /O(s1 + s2)/ size.
+--
+-- @
+-- 'isEmpty'     (overlay x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (overlay x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (overlay x y) >= 'vertexCount' x
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- 'size'        (overlay x y) == 'size' x        + 'size' y
+-- 'vertexCount' (overlay 1 2) == 2
+-- 'edgeCount'   (overlay 1 2) == 0
+-- @
+overlay :: Graph a -> Graph a -> Graph a
+overlay (UG g1) (UG g2) = coerce (G.Overlay g1 g2)
```

Comment:
Should be just `coerce2 G.overlay` I think.

Reply:
Same for some other functions below.

Reply:
I thought so too but had to define and auxiliar function to help GHC:

```Haskell
-- Help GHC with type inference (direct use of 'coerce' does not
-- compile).
coerce3 :: (Coercible a b, Coercible c d) => (a -> c -> G.Graph e) -> (b -> d -> Graph e)
coerce3 = coerce
```

This one can be used instead of `coerce2`!

Reply:
Aha, I see -- perhaps, you should just make `coerce2` more general using your implementation. By `coerceN` above I meant to name specialised `coerce` functions with N arguments.

Reply:
Right now I have 5 coerce type signatures, should I put them somewhere else or can I keep them in the file as of now?

Reply:
Keep them here for now.

Reply:
Okay!